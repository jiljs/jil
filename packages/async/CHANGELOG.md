# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [0.2.4](https://github.com/jiljs/jil/compare/@jil/async@0.2.3...@jil/async@0.2.4) (2022-01-15)

**Note:** Version bump only for package @jil/async





## [0.2.3](https://github.com/jiljs/jil/compare/@jil/async@0.2.2...@jil/async@0.2.3) (2022-01-13)

**Note:** Version bump only for package @jil/async





## [0.2.2](https://github.com/jiljs/jil/compare/@jil/async@0.2.1...@jil/async@0.2.2) (2022-01-13)

**Note:** Version bump only for package @jil/async





## [0.2.1](https://github.com/jiljs/jil/compare/@jil/async@0.2.0...@jil/async@0.2.1) (2022-01-12)

**Note:** Version bump only for package @jil/async





# 0.2.0 (2022-01-10)


### Features

* initial add async module ([3edf673](https://github.com/jiljs/jil/commit/3edf6739c65f0bd8098cd47b3d09fdd96acfa7b7))
