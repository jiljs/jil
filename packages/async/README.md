# @jil/async
> A common async library
