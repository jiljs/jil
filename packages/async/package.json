{
  "name": "@jil/async",
  "description": "A Fibonacci and exponential async library",
  "version": "0.2.4",
  "keywords": [],
  "main": "index.js",
  "types": "index.d.ts",
  "engines": {
    "node": ">=14"
  },
  "author": "TY",
  "license": "MIT",
  "scripts": {
    "version": "npm run copy:assets",
    "copy:assets": "cpy package.json README.md dist",
    "build": "nb-tsc && npm run copy:assets",
    "clean": "nb-clean jil-async*.tgz *.tsbuildinfo package dist/*",
    "pretest": "npm run clean && npm run build",
    "test": "npm run jest",
    "jest": "jest",
    "verify": "npm pack && tar xf jil-async*.tgz && tree package && npm run clean"
  },
  "dependencies": {
    "@jil/cancellation": "^0.2.5",
    "@jil/errors": "^0.2.2",
    "@jil/types": "^0.2.3"
  },
  "devDependencies": {
    "@nutol/build": "^0.2.1",
    "@nutol/eslint-config": "^0.2.0",
    "@types/jest": "^27.4.0",
    "@types/node": "^17.0.8",
    "cpy-cli": "^3.1.1"
  },
  "files": [
    "**/*",
    "!__tests__"
  ],
  "publishConfig": {
    "access": "public",
    "directory": "dist"
  }
}
