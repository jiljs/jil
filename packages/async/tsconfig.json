{
  "$schema": "http://json.schemastore.org/tsconfig",
  "extends": "@nutol/build/config/tsconfig.common.json",
  "compilerOptions": {
    "outDir": "dist",
    "rootDir": "src",
    "composite": true
  },
  "include": [
    "src"
  ],
  "references": [
    {
      "path": "../cancellation/tsconfig.json"
    },
    {
      "path": "../errors/tsconfig.json"
    },
    {
      "path": "../types/tsconfig.json"
    }
  ]
}
