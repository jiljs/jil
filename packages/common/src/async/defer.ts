/* eslint-disable @typescript-eslint/no-explicit-any */
import {canceled} from '../errors/canceled';

export type ResolveFn<T = unknown> = (value?: T | Promise<T>) => void;
export type RejectFn = (reason?: unknown) => void;

export interface DeferredPromise<T> extends Promise<T> {
  readonly isRejected: boolean;
  readonly isResolved: boolean;
  readonly isSettled: boolean;

  /**
   * The deferred promise.
   */
  promise: Promise<T>;

  /**
   * Resolves the promise with a value or the result of another promise.
   * @param value - The value to resolve the promise with.
   */
  resolve(value?: T | Promise<T>): void;
  complete(value?: T | Promise<T>): void;

  /**
   * Reject the promise with a provided reason or error.
   * @param reason - The reason or error to reject the promise with.
   */
  reject(reason?: unknown): void;
  error(reason?: unknown): void;

  /**
   * Reject with canceled error
   */
  cancel(): void;
}

/**
 Create a deferred promise.
 @example
 ```
 function delay(milliseconds) {
	const deferred = defer();
	setTimeout(deferred.resolve, milliseconds, 'ðŸ¦„');
	return deferred.promise;
	// or
	// return deferred;
}
 console.log(await delay(100));
 //=> 'ðŸ¦„'
 ```
 */
export function defer<T>(): DeferredPromise<T> {
  let rejected = false;
  let resolved = false;
  let resolveFn: Function;
  let rejectFn: Function;

  const deferred = <DeferredPromise<T>>new Promise((c, e) => {
    resolveFn = c;
    rejectFn = e;
  });

  const resolve: ResolveFn<T> = value => {
    resolved = true;
    resolveFn(value);
  };

  const reject: RejectFn = reason => {
    rejected = true;
    rejectFn(reason);
  };

  deferred.promise = deferred;

  deferred.resolve = resolve;
  deferred.complete = resolve;
  deferred.reject = reject;
  deferred.error = reject;
  deferred.cancel = () => reject(canceled());

  void Object.defineProperty(deferred, 'isRejected', {
    get(): any {
      return rejected;
    },
  });

  void Object.defineProperty(deferred, 'isResolved', {
    get(): any {
      return resolved;
    },
  });

  void Object.defineProperty(deferred, 'isSettled', {
    get(): any {
      return resolved || rejected;
    },
  });

  return deferred;
}
