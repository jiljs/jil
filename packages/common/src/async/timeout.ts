import {CancellationToken} from '../cancellation';
import {canceled} from '../errors/canceled';
import {CancelablePromise, createCancelablePromise} from './cancelable';

export function timeout(millis: number): CancelablePromise<void>;
export function timeout(millis: number, token: CancellationToken): Promise<void>;
export function timeout(millis: number, token?: CancellationToken): CancelablePromise<void> | Promise<void> {
  if (!token) {
    // eslint-disable-next-line @typescript-eslint/no-shadow
    return createCancelablePromise(token => timeout(millis, token));
  }

  return new Promise((resolve, reject) => {
    const handle = setTimeout(() => {
      disposable();
      resolve();
    }, millis);
    const disposable = token.onCancellationRequested(() => {
      clearTimeout(handle);
      disposable();
      reject(canceled());
    });
  });
}

export function delay(millis: number): CancelablePromise<void>;
export function delay(millis: number, token: CancellationToken): Promise<void>;
export function delay(millis: number, token?: CancellationToken): CancelablePromise<void> | Promise<void> {
  return timeout(millis, token as CancellationToken);
}

export function clearableTimeout(handler: () => void, ms = 0): () => void {
  const timer = setTimeout(handler, ms);
  return () => clearTimeout(timer);
}
