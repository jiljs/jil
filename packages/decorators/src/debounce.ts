import {isMethod} from './helpers/is-method';

/**
 * A method decorator that delays the execution of the class method
 * by the provided time in milliseconds.
 */
export function debounce(delay: number): MethodDecorator {
  return (target, property, descriptor) => {
    if (
      process.env.NODE_ENV !== 'production' &&
      (!isMethod(target, property, descriptor) || !('value' in descriptor && typeof descriptor.value === 'function'))
    ) {
      throw new TypeError(`\`@debounce\` may only be applied to class methods.`);
    }

    // We must use a map as all class instances would share the
    // same timer value otherwise.
    const timers = new WeakMap<Function, NodeJS.Timeout>();

    // Overwrite the value function with a new debounced function
    const func = descriptor.value;

    // @ts-expect-error Override generic
    // eslint-disable-next-line @typescript-eslint/no-shadow
    descriptor.value = function debounce(this: Function, ...args: unknown[]) {
      const timer = timers.get(this);

      if (timer) {
        clearTimeout(timer);
        timers.delete(this);
      }

      timers.set(
        this,
        setTimeout(() => {
          (func as unknown as Function).apply(this, args);
        }, delay),
      );
    };
  };
}
