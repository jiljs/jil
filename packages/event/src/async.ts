import {onUnexpectedError} from '@jil/errors/handler';
import {Emitter, ScopedListener} from './emitter';

export interface AsyncEventCancellationToken {
  isCancellationRequested: boolean;
}

export interface IWaitUntil {
  token: AsyncEventCancellationToken;
  waitUntil(thenable: Promise<unknown>): void;
}

export type IWaitUntilData<T> = Omit<Omit<T, 'waitUntil'>, 'token'>;

export class AsyncEmitter<T extends IWaitUntil> extends Emitter<T> {
  private _asyncDeliveryQueue?: Array<[ScopedListener<T>, IWaitUntilData<T>]>;

  async emitAsync(
    data: IWaitUntilData<T>,
    token: AsyncEventCancellationToken,
    promiseJoin?: (p: Promise<unknown>, listener: Function) => Promise<unknown>,
  ): Promise<void> {
    if (!this._listeners) {
      return;
    }

    if (!this._asyncDeliveryQueue) {
      this._asyncDeliveryQueue = [];
    }

    for (const listener of this._listeners) {
      this._asyncDeliveryQueue.push([listener, data]);
    }

    while (this._asyncDeliveryQueue.length > 0 && !token.isCancellationRequested) {
      // eslint-disable-next-line @typescript-eslint/no-shadow
      const [listener, data] = this._asyncDeliveryQueue.shift()!;
      const thenables: Promise<unknown>[] = [];

      const dataWithWaitUntil = <T>{
        ...data,
        token,
        waitUntil: (p: Promise<unknown>): void => {
          if (Object.isFrozen(thenables)) {
            throw new Error('waitUntil can NOT be called asynchronous');
          }
          if (promiseJoin) {
            p = promiseJoin(p, typeof listener === 'function' ? listener : listener[0]);
          }
          thenables.push(p);
        },
      };

      try {
        if (typeof listener === 'function') {
          listener.call(undefined, dataWithWaitUntil);
        } else {
          listener[0].call(listener[1], dataWithWaitUntil);
        }
      } catch (e) {
        onUnexpectedError(e);
        continue;
      }

      // freeze thenables-collection to enforce sync-calls to
      // wait until and then wait for all thenables to resolve
      Object.freeze(thenables);

      await Promise.allSettled(thenables).then(values => {
        for (const value of values) {
          if (value.status === 'rejected') {
            onUnexpectedError(value.reason);
          }
        }
      });
    }
  }
}
