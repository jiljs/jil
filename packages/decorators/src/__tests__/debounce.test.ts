import {timeout} from '@jil/common/async/timeout';
import {debounce} from '../debounce';

describe('@debounce()', () => {
  class Test {
    @debounce(100)
    log(...args: unknown[]) {
      console.log('Logged!', ...args);
    }
  }

  let spy: jest.SpyInstance;

  beforeEach(() => {
    // jest.useFakeTimers();

    spy = jest.spyOn(console, 'log').mockImplementation();
  });

  afterEach(() => {
    spy.mockRestore();

    // jest.useRealTimers();
  });

  it('errors if applied to a class', () => {
    expect(() => {
      // @ts-expect-error Allow decorator here
      @debounce(100)
      class TestClass {}

      return TestClass;
    }).toThrowErrorMatchingSnapshot();
  });

  it('errors if applied to a property', () => {
    expect(() => {
      class TestProp {
        // @ts-expect-error Allow decorator here
        @debounce(100)
        value = 123;
      }
      return TestProp;
    }).toThrowErrorMatchingSnapshot();
  });

  it('logs only after the delay', async () => {
    const test = new Test();

    test.log();

    expect(spy).not.toHaveBeenCalled();

    test.log();

    expect(spy).not.toHaveBeenCalled();

    // jest.advanceTimersByTime(150);
    await timeout(100);

    expect(spy).toHaveBeenCalledWith('Logged!');
    expect(spy).toHaveBeenCalledTimes(1);
  });

  it('passes arguments through', async () => {
    const test = new Test();

    test.log(1, 2);
    test.log(3, 4);
    test.log(5, 6);

    // jest.advanceTimersByTime(150);
    await timeout(100);

    expect(spy).toHaveBeenCalledWith('Logged!', 5, 6);
    expect(spy).toHaveBeenCalledTimes(1);
  });
});
