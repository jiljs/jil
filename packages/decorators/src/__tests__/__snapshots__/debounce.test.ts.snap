// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@debounce() errors if applied to a class 1`] = `"\`@debounce\` may only be applied to class methods."`;

exports[`@debounce() errors if applied to a property 1`] = `"\`@debounce\` may only be applied to class methods."`;
